design/user-profile.md

```diff
@@ -86,27 +86,63 @@ A very rough idea on how the user profile JSON could look like:
     {
         "attributes": [{
             "name": "department",
-            "view": ["admin", "user"],
-            "edit": ["admin"],
+            "permissions": {
+                    "view": ["admin", "user"], 
+                    "edit": ["admin"],
+            }
             "requirement":"always",
             "validation": {
                 "name", 
                 { "length" : { "min" : 10, "max": 20 } }
+            },
+            "processors": {
+                "datetime-converter": {
+                    "datetime-format" : "tt/mm/yyyy"
+                },
+                "timezone-converter": {
+                    "z": "UTC-0"
+                };
+            },
+            "annotations":  {
+                "key": "value",  
+                "framecolor": "red",
+                "gui-order": "1",
+                "type": "dropdown",
+                "defaults" : [1,2,3,5]
             }
-        }
+        }]
     }
 
+### Required Fields
+
 Requirement should support following values:
 
 * `optional` (default if no value given)
 * `always`
 * `user`
-* `{ "scope" : "scope-1" }`
-* `{ "scope" : [ "scope-1", "scope-2" ] }`
 
 The `user` requirement allows an admin to create a user without specifying the value, while the user will then be required to enter the value on first login.
 
-The `scope` requirement allows an attribute to only be required when a specific client scope is being requested.
+### Profile Attribute Requirement Configuration SPI
```

Comment:
Can you elaborate on other use-cases than client scope here? I would like to understand the need for other use-cases and an SPI before we bring in that additional level of complexity.

Based on the example I don't understand how this would be applied to the user profile. Where does the snippet go in the profile JSON? Before we had a simple requirement field attached to each attribute.

Reply:
Regards the profile JSON. The idea is to store the configuration with the element which defines it. E.g. configs for clients are defined on the client level and are also stored on the client level. 

The idea is to support any kind of profile override in a given context. Example: a custom required action would like to add mandatory fields for the profile. Then it would implement this interface and configures the fields in its config. When a user tries to use the action he will be prompted with the merged configuration. 

In general, the idea is to establish a general configuration manipulation hook for all contexts which can reasonably try to enrich the profile based on their state. For example, imagine different mandatory fields are given if you are logging in via different idps. If you log in from BusinessContextA you want other mandatory fields than in BusinessContextB but it's not decided on client level but on the login context. The idea is to support any kind of these use cases with an extensional API.

Reply:
Afraid you've completely lost me here. I don't understand how this is supposed to be configured/used.

From my perspective a client needs a specific client scope, and a client scope requires the user profile to contain specific attributes.

I don't see a use-case for anything beyond that. Can you explain the use-cases you are trying to cover, and also some more details on how this is configured/used?

Reply:
If you abstract the problem away from scopes at all. And remove the coupling to scopes for a second it may get more clear. 

The scope topic in general terms tries to achieve manipulation of the profile configuration based on scope values. 
Therefore we need to process "some" code in case a flow happens in a specific context (a client in this case). Instead of hardcoding the scope handling, we propose to hardcode the ways the profile can be manipulated.

For example, a simple interface for the SPI could look like this
`Set<String> getMandatoryAttributes()`

This interface is called when the profile (which always gets the session context) determines that the SPI is registred for this contexts (e.g. for an authentication session for a specific client) 

This enables us to define any kind of SPI to manipulate the mandatory fields in the profile in different contexts. For example, an Identity provider implementation could set mandatory fields based on the provider config. So different providers would enforce different fields for the same user (e.g. to reflect a specific business context, or fields which are automatically set in on case but not in the other) 

The idea is to introduce a general notion of how other components than the profile are able to manipulate the config based on their context. 

I hope this makes it more explicit. If you feel that there is no general need to enforce different profile behavior based on different contexts (provider, client, account ui, resource api) we can drop this whole approach for sure.

Reply:
One core requirement here is the ability to have progressive user profiles. The idea on how to solve that would be that users can register with an initial basic set of attributes, but then as clients need more information about the user the users would be required to provide this information. This use-case is well solved by only requiring some attributes for specific client scopes, as now clients can request a default client scope to always require additional attributes, or can have optional scopes to require it only when they need it. Beyond that I don't think there's a case to configure different attribute requirements for a client.

Then there's the case of an external user authenticating which is a lot more complex. An LDAP provider may for example not have all the attributes required by a realm. What happens then? If the LDAP provider is read-only and doesn't allow storing additional things in KC DB, what do we do? Same goes to IdPs once we enable transient users (users only exists in KC during the time they are logged-in, but are not persisted to the KC DB). I can think of a lot of clever things we can do in these cases - including the ability to validate a LDAP configuration based on the user profile to make sure that all required attributes are mapped, including having drop-downs when selecting user attributes rather than free-text. We can also have options to reject logins if they don't match the required profile, ask the user to update the profile, or just ignoring it.

I can also see the use-case of having different attributes for different groups of users. For example you may require employees to provide an employment-id, while customers don't have to. This could be based on marking attributes as required if a user has a attribute/role/group.

I don't think it's a good idea to have an IdP implementation able to alter the user attributes, that would probably quickly become unmanageable and a bit of a spaghetti code mess. From my perspective the user profile is something that is defined based on requirements on user groups as well as what clients need, as such it's rather the user profile that instructs the IdP or user storage provider, not the other way around.

Reply:
As we also see your concerns we will go with the client only approach as discussed in the initial draft. We will roll back the changes in the PR soon.